### PR TITLE
Defense channels and rolemapping

### PR DESCRIPTION
**Pull request by zuuring**
### Notes
This adds a new discord command `defense-audit` which has one parameter `client-name` which will then perform the following steps:

- Create two channels, `int-clientname-audit` and `ext-clientname-audit` 
- Create two new roles `Defense Internal: clientname` and `Defense Extenal: clientname`
- Grant `ViewChannel` access to both channels for `Defense Internal: *` and only `ext-*-audit` channel access to `Defense Extenal`
- Generate an invite link for both channels

### Screenshot
Screenshot 2024-03-20 at 13 51 15

### Automatic role assignment
Also included in this PR is automatic role assignment based of the invite code used. Due to lack of IDs associated to `GuildMemberAdd` events when a new user joins discord, the only way to tie a discord user to an invite link is to count invite code uses within the guild and compare it once a new member has joined the discord. For this to work, we check the invites and match role assignment to channel.name the invite link was generated from. 

### To-do
- [x] Setup `defense-audit` command
- [x] Add rolemapping based on invite joins
- [x] Fix issue on `int-client-name-audit` channels not finding correct role for `ext-client-name-audit` channels

### CODE REVIEW

File: discord-scripts/invite-management.ts [hunk 1]
```diff
@@ -133,10 +209,86 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+        const channel = member.guild.channels.cache.get(
+          usedInvite.channelId,
+        ) as TextChannel
+        if (channel) {
+          const channelTypeMatch = channel.name.match(/(ext|int)-(.*)-audit/)
+          const clientName = channelTypeMatch
+            ? channelTypeMatch[2]
+                .replace(/-/g, " ")
+                .split(" ")
+                .map(
```
Comment (zuuring) [inline discord-scripts/invite-management.ts:260]: Going to refactor this around to just match ChannelID -> ChannelName -> Assigned roles and then grab role with `Defense External: Client` rather than falling overmyself with these consts 😄

File: discord-scripts/invite-management.ts [hunk 2]
```diff
@@ -133,10 +209,86 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+          string,
+          { uses: number }
+        >) || new Collection<string, { uses: number }>()
+      const fetchedInvites = await member.guild.invites.fetch()
+      const newInvites = new Collection<string, number>(
+        fetchedInvites.map((invite) => [invite.code, invite.uses ?? 0]),
+      )
+      guildInvites.set(member.guild.id, newInvites)
+
+      robot.logger.info(
```
Comment (zuuring) [inline discord-scripts/invite-management.ts:234]: Will streamline this to reduce loggin for debugging and fix up `oldInvites` and `newInvites`

**Comment by Shadowfiend** [comment]
Question here: can we kill the embeds in the intro message?

File: discord-scripts/invite-management.ts [hunk 3]
```diff
@@ -1,8 +1,15 @@
...
+  Client,
+  ChannelType,
+  Collection,
+  TextChannel,
+  GuildMember,
+} from "discord.js"
 import { DAY, MILLISECOND, WEEK } from "../lib/globals.ts"
 
-const EXTERNAL_AUDIT_CHANNEL_REGEXP = /^ext-(?<client>.*)-audit$/
+const guildInvites = new Collection()
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:12]: Any reason not to just use objects here?

File: discord-scripts/invite-management.ts [hunk 4]
```diff
@@ -22,10 +29,34 @@ async function createInvite(
   }
 }
 
+async function listInvites(discordClient: Client, robot: Robot): Promise<void> {
+  discordClient.guilds.cache.forEach(async (guild) => {
+    const fetchInvites = await guild.invites.fetch().catch((error) => {
+      robot.logger.error(
+        `Failed to fetch invites for guild ${guild.name}: ${error}`,
+      )
+    })
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:38]: When we use `async`/`await`, let's use `try`/`catch` rather than `.catch` behavior. It'll clean up the control flow in this whole function.

File: discord-scripts/invite-management.ts [hunk 5]
```diff
@@ -38,6 +69,28 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+      (command) => command.name === "defense-audit",
+    )
+    if (existingDefenseCommand === undefined) {
+      robot.logger.info("No defense-audit command found, creating it!")
+      await application.commands.create({
+        name: "defense-audit",
+        description: "Creates Defense audit channels",
+        options: [
+          {
+            name: "client-name",
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:84]: Let's call this `audit-name`.

File: discord-scripts/invite-management.ts [hunk 6]
```diff
@@ -38,6 +69,28 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+            type: ApplicationCommandOptionType.String,
+            description: "The name of the client to create the channels for",
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:86]: Let's say “The name of the audit/client to create the channels for.”

File: discord-scripts/invite-management.ts [hunk 7]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+        })
+        return
+      }
 
-            const role = await channel.guild.roles.create({
-              name: roleName,
-              reason: `Role for ${channel.name} channel`,
-            })
+      if (
+        !interaction.guild ||
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:149]: Can't happen after the previous statement, right?

File: discord-scripts/invite-management.ts [hunk 8]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+        !(interaction.channel instanceof TextChannel) ||
+        (interaction.channel.parent &&
+          interaction.channel.parent.name !== "defense")
+      ) {
+        await interaction.reply({
+          content:
+            "This command can only be run in channels under the 'defense' category.",
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:156]:
```suggestion
            "This command can only be run in chat channels under the 'Defense' category.",
```

File: discord-scripts/invite-management.ts [hunk 9]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
-            await channel.permissionOverwrites.create(role, {
-              ViewChannel: true,
+      const clientName = interaction.options.get("client-name")
+      if (!clientName) {
+        await interaction.reply({
+          content: "Client name is required for the defense-audit command.",
+          ephemeral: true,
+        })
+        return
+      }
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:169]: To be updated based on the above comment as well.

File: discord-scripts/invite-management.ts [hunk 10]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+
+      try {
+        if (typeof clientName.value === "string") {
+          await interaction.deferReply({ ephemeral: true })
+          const normalizedClientName = clientName.value
+            .replace(/\s+/g, "-")
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:175]: Do we have any other limitations on names here? e.g. special characters. Wonder if this is more like `.replace(/[^a-zA-Z0-9]/g, "-")`.

File: discord-scripts/invite-management.ts [hunk 11]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+            .toLowerCase()
+          const internalChannelName = `int-${normalizedClientName}-audit`
+          const externalChannelName = `ext-${normalizedClientName}-audit`
+
+          const defenseCategory = interaction.guild.channels.cache.find(
+            (c) => c.name === "defense",
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:181]: `c.type` should also be a category, right? I don't think we have a plain `#defense` channel, but I can imagine a scenario where we would.

File: discord-scripts/invite-management.ts [hunk 12]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+            name: internalChannelName,
+            type: ChannelType.GuildText,
+            parent: defenseCategory.id,
+          })
+
+          const externalChannel = await interaction.guild.channels.create({
+            name: externalChannelName,
+            type: ChannelType.GuildText,
+            parent: defenseCategory.id,
+          })
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:202]: Let's check if these exist before creating—i.e. do more of a `createOrUpdate` operation where we create it if it exists, and if it doesn't then we sync permissions.

File: discord-scripts/invite-management.ts [hunk 13]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+          const internalRoleName = `Defense Internal: ${clientName.value}`
+          const externalRoleName = `Defense External: ${clientName.value}`
+          const internalRole = await interaction.guild.roles.create({
+            name: internalRoleName,
+            reason: "Role for internal audit channel",
+          })
+          const externalRole = await interaction.guild.roles.create({
+            name: externalRoleName,
+            reason: "Role for external audit channel",
+          })
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:213]: Same deal here—though maybe `.create` on the Discord API already does this?

File: discord-scripts/invite-management.ts [hunk 14]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+          await externalChannel.permissionOverwrites.create(externalRole, {
+            ViewChannel: true,
+          })
+          await externalChannel.permissionOverwrites.create(internalRole, {
+            ViewChannel: true,
+          })
+
+          const internalInvite = await createInvite(internalChannel)
+          const externalInvite = await createInvite(externalChannel)
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:226]: For clarity, I would do all of the internal channel stuff first, and then all of the external channel stuff. That'll make it easier to tell what we're doing to which channel, and identify if we've messed up in the future.

File: discord-scripts/invite-management.ts [hunk 15]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+
+          await interaction.editReply({
+            content:
+              `**Defense audit setup complete for: ${clientName.value}**\n\n` +
+              `Internal Channel: <#${internalChannel.id}> (Invite: ${internalInvite.url})\n` +
+              `External Channel: <#${externalChannel.id}> (Invite: ${externalInvite.url})\n\n` +
+              `Roles created: <@&${internalRole.id}>, <@&${externalRole.id}>`,
+          })
+        }
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:235]: We should still send the initial message to the internal and external channels that we were sending before—it'll help to have a reference point inside the channels with the invite/permissions info.

File: discord-scripts/invite-management.ts [hunk 16]
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+                .replace(/-/g, " ")
+                .split(" ")
+                .map(
+                  (word) =>
+                    word.charAt(0).toUpperCase() + word.slice(1).toLowerCase(),
+                )
+                .join(" ")
+            : ""
+
+          robot.logger.info(`Channel Name: ${channel.name}`)
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:284]: If we keep this log we should give it context as to what process is logging it.

**Review by Shadowfiend — CHANGES_REQUESTED** [review]
Left a handful of questions & comments.

**Comment by zuuring** [comment]
@Shadowfiend Just pushed an update to this with fixes to the review comments as following: 
- Update permissions if channel already exists
- Check if role already exists, if so skip
- Changed logic so that if audit channels already exist, we output a different message to interaction
- Remove embeds by wrapping invite URL in quotes.
- Change verbiage around messages
- Separates InternalChannel and externalChannel
- Changed invite code permission match

Now one thing that came up, turns out Discord channel categories have a limit of max (50) channels per category. Should we maybe setup an "Audit" category or even (not ideal) a channel category on a per client basis?

File: discord-scripts/invite-management.ts (hunk 16 above)
Comment (zuuring) [reply discord-scripts/invite-management.ts]: Added a better logger at the end of this, so we see when invite is used + role matched to.

File: discord-scripts/invite-management.ts (hunk 15 above)
Comment (zuuring) [reply discord-scripts/invite-management.ts]: Setup a @here message to be sent to the channels being setup, we could also tag the initiated user if needed.

File: discord-scripts/invite-management.ts (hunk 13 above)
Comment (zuuring) [reply discord-scripts/invite-management.ts]: Switched this one around as it seems like actually discord `.create` does not, it will make an ever infinite amount of roles with the same name! Also the case for channel names, I am seeing 😆

**Comment by Shadowfiend** [comment]
> Now one thing that came up, turns out Discord channel categories have a limit of max (50) channels per category. Should we maybe setup an "Audit" category or even (not ideal) a channel category on a per client basis?

Lawd. I think we can hold for now, as I don't think we're going to have 20 audits going in parallel, and older audits will get archived. Making them their own category is a good idea though and we can discuss with Defense once we've got the basics working here.

File: discord-scripts/invite-management.ts [hunk 17]
```diff
@@ -31,19 +30,19 @@ async function createInvite(
 
 async function listInvites(discordClient: Client, robot: Robot): Promise<void> {
   discordClient.guilds.cache.forEach(async (guild) => {
-    const fetchInvites = await guild.invites.fetch().catch((error) => {
+    try {
+      const fetchInvites = await guild.invites.fetch()
+      if (fetchInvites) {
+        guildInvites[guild.id] = guildInvites[guild.id] || {}
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:36]: Fancy-shmancy here is `guildInvites[guild.id] ??= {}`.

File: discord-scripts/invite-management.ts [hunk 18]
```diff
@@ -172,13 +172,13 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
         if (typeof clientName.value === "string") {
           await interaction.deferReply({ ephemeral: true })
           const normalizedClientName = clientName.value
-            .replace(/\s+/g, "-")
+            .replace(/[^a-zA-Z0-9]/g, "-")
             .toLowerCase()
-          const internalChannelName = `int-${normalizedClientName}-audit`
-          const externalChannelName = `ext-${normalizedClientName}-audit`
+          const internalChannelName = `🔒int-${normalizedClientName}-audit`
+          const externalChannelName = `🔒ext-${normalizedClientName}-audit`
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:178]: IMO we should let the existing automation handle this, unless it's not working for some reason.

File: discord-scripts/invite-management.ts [hunk 19]
```diff
@@ -189,49 +189,98 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
-            name: externalChannelName,
-            type: ChannelType.GuildText,
-            parent: defenseCategory.id,
-          })
+          // Internal channel setup
+          let internalChannel = interaction.guild.channels.cache.find(
+            (channel) => channel.name === internalChannelName,
+          ) as TextChannel
+          const internalChannelCreated = !internalChannel
+          if (internalChannelCreated) {
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:197]: This should be `internalChannelNeedsCreation` or something, right?

File: discord-scripts/invite-management.ts [hunk 20]
```diff
@@ -189,49 +189,98 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
...
+              reason: "Role for internal audit channel",
+            })
+          }
 
+          if (internalChannel) {
+            await internalChannel.permissionOverwrites.create(internalRole, {
+              ViewChannel: true,
+            })
+            await internalChannel.send(
+              `@here **Welcome to the ${clientName.value} Internal Audit Channel!**`,
```
Comment (Shadowfiend) [inline discord-scripts/invite-management.ts:221]: Sorry, what I meant with this message is that we should keep the full original version. i.e.:

```
**Welcome to the ${clientName.value} audit internal channel!**

You can use this invite to ...: ${internalInvite.url}.
```

And similar for external except it would say “Welcome to the ${clientName.value} audit channel...” (note that it doesn't say “external”, it just doesn't say “internal”).

The goal here is to have a point (possibly pinned) in each channel where folks can:
- See the invite code.
- See information about the invite code (e.g. when it expires and max uses).

Basically the same stuff we were showing @ https://github.com/thesis/valkyrie/pull/301/files#diff-b34d961b5563321a08dc8493fd258c4f5cec44c90e8edd92999c529b43b0f7aaL87-L98 .

**Review by Shadowfiend — CHANGES_REQUESTED** [review]
Think I'm going to hold this up on the remaining remarks—it's close though!

File: discord-scripts/invite-management.ts (hunk 18 above)
Comment (zuuring) [reply discord-scripts/invite-management.ts:178]: This was a weird one where it seemed to not work like it should of without this definition and 🔒 prefix.

File: discord-scripts/invite-management.ts (hunk 20 above)
Comment (zuuring) [reply discord-scripts/invite-management.ts:221]: That makes sense and will get this changed over! Only question I have about that, if we pin the invite information to each channel, should we worry about deleting that message once a client joins or just simply always have it visible, and just assume team members can remove message if needed. I love the expiry / uses idea, that makes a lot of sense and also helps track.

Another layer, we could even update pinned message with a `@user` of who has claimed invite? If it seems important to surface.

**Comment by zuuring** [comment]
@Shadowfiend Just pushed an update on this with the fixes based on your review! I also added a way to show expiry date using Discord's built in time rendering. That way we get a nice hover event when going over the expiry time.
Screenshot 2024-04-22 at 16 06 29

File: discord-scripts/invite-management.ts (hunk 20 above)
Comment (Shadowfiend) [reply discord-scripts/invite-management.ts]: Love the idea of updating users & counts, though I would say not top priority.

**Review by Shadowfiend — APPROVED** [review]
Changes look good, let's :shipit:
